test/e2e/config/docker.yaml

```diff
@@ -347,7 +347,7 @@ intervals:
   default/wait-control-plane: ["10m", "10s"]
```

Comment:
The first of these log lines is at "Wednesday, March 13, 2024 5:04:51.371 PM" so 20m before the timeout

Reply:
Based on the code we only get to this log line if we successfully called delete on the KCP object. But we don't get a deletionTimestamp (?)

Reply:
Just found this in the KCP controller logs

> {"ts":1710349491885.686,"caller":"controller/controller.go:329","msg":"Reconciler error","controller":"kubeadmcontrolplane","controllerGroup":"controlplane.cluster.x-k8s.io","controllerKind":"KubeadmControlPlane","KubeadmControlPlane":{"name":"quick-start-tw0aec-xs5p6","namespace":"quick-start-djy4ka"},"namespace":"quick-start-djy4ka","name":"quick-start-tw0aec-xs5p6","reconcileID":"cf9d2c37-5e2e-429b-84f7-e81c7b399812","err":"no kind is registered for the type v1beta1.MachinePoolList in scheme \"pkg/runtime/scheme.go:100\""}

> https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/10141/pull-cluster-api-e2e-blocking-main/1767956869067837440/artifacts/clusters/bootstrap/logs/capi-kubeadm-control-plane-system/capi-kubeadm-control-plane-controller-manager/capi-kubeadm-control-plane-controller-manager-979655789-stct7/manager.log

Reply:
Okay found the issue.

I guess to our all surprise the KCP controller has MachinePool specific code, even though we never exposed the feature gate on the command line in our manager.yaml. By changing the default value in features.go we enabled this code (I guess for the first time). That is also why it wasn't enabled via EXP_MACHINE_POOL in docker.yaml

https://github.com/kubernetes-sigs/cluster-api/blob/1c760a5953cc9af2a78a51b943eb19dc9ead0a33/controlplane/kubeadm/internal/controllers/controller.go#L540-L545

Reply:
Pushed a commit to fix it.

This basically means that the KCP controller never waited until all MachinePools were gone before deleting control plane machines. I think this was okay because the Cluster controller never triggered KCP deletion before all MachinePools / MDs were gone.


cc @fabriziopandini @chrischdi (just fyi)
cc @willie-yao (just in case this explains any issues in MP tests or MPs in general)

Reply:
Seems like it might not yet be working; e2e looks like it's going to time out. (I'll remove the 20m timeout I added that of course didn't help.)

Reply:
@mboersma looks like we also have to fix rbac https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/10141/pull-cluster-api-e2e-blocking-main/1768301498602622976/artifacts/clusters/bootstrap/logs/capi-kubeadm-control-plane-system/capi-kubeadm-control-plane-controller-manager/capi-kubeadm-control-plane-controller-manager-7c8c77b6b7-6b8bn/manager.log